### PR TITLE
Hoist style rules to top of the module

### LINKED ISSUE #289: Hoist style rules to the top of the module
Move each stylesheet declaration to the top of the module.

If a stylesheet has already been found - do nothing - but reference the identifier in the ast.

See: https://github.com/atlassian-labs/compiled/pull/274/files?file-filters[]=.json&file-filters[]=.lock&file-filters[]=.tsx&hide-deleted-files=true#diff-f497fa51195bedd5e35d547e1401d219R229

Each declaration should be given a unique, short name. Start with a and work your way up. You should be able to use the inbuilt utils in Babel to get a unique name.

### PR DESCRIPTION
**Pull request by madou**
Closes #289

### CODE REVIEW

**Comment by pgmanutd** [comment]
Looking into it Michael 👀. Give me sometime 😄

**Comment by madou** [comment]
@pgmanutd after we merge my two open PRs I think theres an opportunity to improve the function APIs for ast-builder - they're a little.. wild atm.

Thinking to type them something like:

```js
function compiledComponent(node: t.Node, css: CSSOutput, meta: Metadata) {}
function styledComponent(tag: Tag, css: CSSOutput, meta: Metadata) {}
```

File: packages/babel-plugin/src/styled/index.tsx [hunk 1]
```diff
@@ -124,12 +124,12 @@ export const visitStyledPath = (
   const cssOutput = buildCss(styledData.cssNode, meta);
 
   path.replaceWith(
-    buildStyledComponent({
-      ...meta.state.opts,
-      cssOutput,
-      tag: styledData.tag,
-      parentPath: path as NodePath,
-      scope: path.scope,
```
Comment (pgmanutd) [inline packages/babel-plugin/src/styled/index.tsx:132]: Are we not passing updated `parentPath` and `scope` now 🤔?

File: packages/babel-plugin/src/utils/ast-builders.tsx [hunk 2]
```diff
@@ -1,43 +1,38 @@
...
 import { transformCss } from '@compiled/css';
 import isPropValid from '@emotion/is-prop-valid';
-
-import { PluginOptions, Tag } from '../types';
-
+import { Tag } from '../types';
 import { CSSOutput } from './css-builders';
 import { pickFunctionBody } from './ast';
+import { CompiledOpts, CompiledTemplateOpts, StyledOpts, StyledTemplateOpts } from './types';
+import { Metadata } from '../types';
```
Comment (pgmanutd) [inline packages/babel-plugin/src/utils/ast-builders.tsx:11]: nit: Two `types` files confused me at first 😅? Can we do something about these 🤔?

File: packages/babel-plugin/src/styled/index.tsx (hunk 1 above)
Comment (madou) [reply packages/babel-plugin/src/styled/index.tsx:132]: we are just through metadata instead

File: packages/babel-plugin/src/utils/ast-builders.tsx (hunk 2 above)
Comment (madou) [reply packages/babel-plugin/src/utils/ast-builders.tsx:11]: oops

any suggestions? main thing i wanted to improve was the size of types is getting a bit crazy atm

File: packages/babel-plugin/src/styled/index.tsx (hunk 1 above)
Comment (pgmanutd) [reply packages/babel-plugin/src/styled/index.tsx:132]: hmmm I just thought we are passing updated `parentPath` and `scope` separately. No worries, if we are now passing these through meta. This new api looks more clean 🎉

File: packages/babel-plugin/src/utils/ast-builders.tsx (hunk 2 above)
Comment (pgmanutd) [reply packages/babel-plugin/src/utils/ast-builders.tsx:11]: Yeah they do. `{ CompiledOpts, CompiledTemplateOpts, StyledOpts, StyledTemplateOpts }` seems more like `ASTUtilsTypes` something. Maybe we can have something better naming than this.

Comment (pgmanutd) [reply packages/babel-plugin/src/utils/ast-builders.tsx:11]: Are we making this change @madou? Otherwise this looks good.

Comment (madou) [reply packages/babel-plugin/src/utils/ast-builders.tsx:11]: let's make the change in a follow up PR 👍 (will modify the APIs of above funcs too)

**Review by pgmanutd — APPROVED** [review]
lgtm

**Pull request merged**